### DISCUSSION TITLE
[Question] Is it possible to calc with variables in PDF [part payment invoice]?

### DISCUSSION THREAD

**Discussion by mad99** [YetiForceCompany/YetiForceCRM#11467] [How to]
Hi again,
Ca't find any hint about:

I would like to create a invoice with «part payment».

### Does anyone have any experience with «part payment invoice»?
Our customers can pay in three parts.

So I have to create 3 invoices.
First 33% of full amount.
Second 33% of full amount.
Third 34% of full amount.

Is there a way to calc with variables? 
Something like $(record : sum_gross) * 0.33$

#### Or is there an other known approach for create a «part payment invoice»?
Something like a form-field for «total pay for this invoice»?

Would be great to get some hints or links about.
Regards & thank you
Martin

**Comment by davide-alghi** [comment]
Hi mad99,
there are many ways to manage installments payment.
I suggest you to produce only one sales invoice and send periodically an email to the customer with the relative payment request details of single part / installment: if you provide 3 installments per invoice and you produce 4000 invoice / year, you'll have 12000 invoice to manage; your costs will increase and for your customer too.
**So I assume to produce a single sales invoice with a maximum of 5 installments / parts.**

**Payment in** module help you to trace payments and, automatically, detect waiting / partially /paid sales invoices. How to? Following ...

### Managing of payments in
1. into **Payment in** module add a new custom field named **Sales invoice**: type **relation (1:m)**, with **Sales invoice** as target (now you can _assign_ single payment into sales invoice and in sales invoice you can list all payments in) (NB if you are running with YF 5.2, there already exist this relation field)
2. into **Sales invoice** module add a new custom field named **Paid amount**: type **currency**
3. create a workflow on **payment in** module, with task to add payment-in.amount field to sales-invoice.paid_amount field (into Paid amount you'll have paid total)
4. create 3 workflows on **Sales invoice** module to update Payment status: waiting / partially / paid; based on **Paid amount** and **Gross amount** fields

### Managing of notification via email
Only a guideline: in **Sales invoice** module, you could create a new block for each installment / part, with 3 fields: amount to pay, date to send the notification (email), a flag for sent status (yes / no).
After this you can create a workflow for each part (block) to send email and set installment / part block

**Comment by mad99** [comment]
Wouw thank you @davide-alghi,
It sound a bit overwhelming. Do you have a good link to learn more about all the YetiForce stuff, like «fields», «workflow» and «module-extending»?

I guess, a part-payment could add into YetiForce with «Accounting» > «Payments in»? Right?

Your solution looks great, but there are a lot of unfamiliar part in.
I see your perspective with 4000/12000 invoices. But this would be a small disadvantage. I could do all what I need with the knowing I have ;o))
At the end, it is easier for us and our customers to have each part of the business-cooperation apart. So the 3 Invoices are not to bad for both and easy to use.
The more work, I will change with a usable and simple workflow in YetiForce.

I only have to find a solution to add items from «Payments in» to a new invoice.
Maybe you know a possibility to use the «Payments in» in a new invoice (with own Class in TextParser)?

Maybe I can figure out this/make a work around with the «Correcting Invoice». Maybe there is some query in to get the deposit in a new invoice.
Or maybe I can create an «Correcting Invoice» with a new Invoice-Number and the deposit on it.
Then all would be fine.

But unfortunately in «Correcting Invoice» there seems to be a bug (https://github.com/YetiForceCompany/YetiForceCRM/issues/11483).

Maybe You have some other idea how I could use the deposit-query for work in an invoice...

Thank you very much for all the information above.
Before I can make use of it, I have learn some basics and read docs.
 If you have some recommendation I would be glad.

Regards Martin

**Comment by davide-alghi** [comment]
Hi Martin, my replies inline ...

> Wouw thank you @davide-alghi,
> It sound a bit overwhelming. Do you have a good link to learn more about all the YetiForce stuff, like «fields», «workflow» and «module-extending»?
> 

The available documentation is here on github (issues) or on YF website.


> I guess, a part-payment could add into YetiForce with «Accounting» > «Payments in»? Right?
> 

Payments in is simply a module to trace payments: you can use it to trace part-payment

> Your solution looks great, but there are a lot of unfamiliar part in.
> I see your perspective with 4000/12000 invoices. But this would be a small disadvantage. I could do all what I need with the knowing I have ;o))
> At the end, it is easier for us and our customers to have each part of the business-cooperation apart. So the 3 Invoices are not to bad for both and easy to use.
> The more work, I will change with a usable and simple workflow in YetiForce.
> 
> I only have to find a solution to add items from «Payments in» to a new invoice.
> Maybe you know a possibility to use the «Payments in» in a new invoice (with own Class in TextParser)?
> 
> Maybe I can figure out this/make a work around with the «Correcting Invoice». Maybe there is some query in to get the deposit in a new invoice.
> Or maybe I can create an «Correcting Invoice» with a new Invoice-Number and the deposit on it.
> Then all would be fine.
> 
> But unfortunately in «Correcting Invoice» there seems to be a bug (#11483).
> 

At this point I no longer understand what your needs are.

> Maybe You have some other idea how I could use the deposit-query for work in an invoice...
> 
> Thank you very much for all the information above.
> Before I can make use of it, I have learn some basics and read docs.
> If you have some recommendation I would be glad.
> 
> Regards Martin

**Comment by mad99** [comment]
Hey @davide-alghi,
Sorry for my difficult explanation.

In simple words but longer sentence:
### Customer-Relationshiop
I'm working with clients on the relationship: German term «Werkvertrag» (English 1:1 Translation: **"contract for work."**
In German the «Werkvertrag» regulates all sales matters that define not «simply products» but **«goods to be produced».**

### Challenge
The relationship is a long-term relationship and for the completion of the work, the customer must also contribute intermediate work.
Not infrequently, some steps, on which my work is based, ***are delayed*** on his side.

### Solution
The currently working solution is a partial invoicing at certain times of cooperation.
- This keeps the motivation on the part of the customer awake, 
- we have much fewer delays and 
- we do not sit on a heap of work without payment.

In Germany/Switzerland/Austria's Contract Legislation there is the term «Abschlagszahlung» (maybe part-payment in english 1:1 translation) The law clearly regulates this and also the specification of what must be contained in it (e.g. a clear tax calculation for each partial invoice)

It works fine and so we are used to break the total amount in 3 Parts 
- 2 of «Abschlagsrechnungen» (Part-invoicing) and at 
- 1 end the Final Invoice.

Simple and best practice with the government.

### Problem
- We have to create 3 different invoices with particular tax on each of them.
- In the second part-invoice, the first part-payment should listed and be off the total.
- In the end-bill we have to add possible services which the customer has ordered additionally and list the 2 part-payments and should also off the total.

All with particular tax for each part-invoice.

### Possible Approach
- For the first invoice, I found a solution with an own coded TextParser: works fine! *(Thanks to the search in the source code and with your information about TextParser I am sure to have found the right place.)*
- For the second and end-bill I found the «AmountToReturn.php» TextParser Class. But unfortunately I can't use it in regular invoices. This restriction probably also makes sense, because the down payment is linked to an existing invoice.
- My hope is to **write also a particular TextParser** for 
-- «correcting invoice» for 2. Part Invoice, and then 
-- for the end-bill.
- I guess, if the correcting-invoice is possible to arrange like a «standard looking» invoice I got the clou/I found a working solution for the need.

### Conclusion
That's what I wanted to write in the last message in a nutshell. But since I am not a native English speaker, those cuts are sometimes unnecessarily complicated or impossible to understand.

Sorry.

- Hope bug will fixed soon. 
- Or my wrong actions that lead to an unexpected error will be improved. I will try different test-invoices today, maybe the "correcting-invoice" works with one of them.
**Than I can try to use an own TextParser-Class.**

### Other Ideas are very welcome!
But maybe you or someone else have a much better idea.
I'm new to YetiForce and totally happy about every thought of experienced users.

Regards
Martin
